### PR TITLE
fix middleware gen bugs & add tests

### PR DESCRIPTION
**Pull request by cl1337**
- fix middleware gen cast to `[]map[string]interface{}`
- fix trans middleware not added to runtime middlewares but leaves an empty slot
- add trans middleware examples
- fix tests

todo: add one e2e in test_case.json, verify req/res especially for override / default init case

cc @uber/zanzibar-team

### CODE REVIEW

File: codegen/gateway.go [hunk 1]
```diff
@@ -627,83 +627,21 @@ func augmentHTTPEndpointSpec(
...
-					fieldMap[toField] = FieldMapperEntry{
-						QualifiedName: fromField,
-					}
-				}
+			resTransforms, err := setTransformMiddleware(middlewareObj)
+			if err != nil {
+				return nil, err
 			}
-			espec.ReqTransforms = fieldMap
+			espec.RespTransforms = resTransforms
```
Comment (cl1337) [inline codegen/gateway.go]: fix set req to resp transform

File: codegen/gateway.go [hunk 2]
```diff
@@ -749,12 +687,12 @@ func augmentHTTPEndpointSpec(
 			}
 		}
 
-		middlewares[idx] = MiddlewareSpec{
+		middlewares = append(middlewares, MiddlewareSpec{
```
Comment (cl1337) [inline codegen/gateway.go]: fix leaving middleware space holder generate empty middleware with default name = ""

File: codegen/gateway.go [hunk 3]
```diff
@@ -825,6 +763,44 @@ func augmentHTTPEndpointSpec(
...
 
+func setTransformMiddleware(middlewareObj map[string]interface{}) (map[string]FieldMapperEntry, error) {
+	fieldMap := make(map[string]FieldMapperEntry)
+	opts, ok := middlewareObj["options"].(map[string]interface{})
+	if !ok {
+		return nil, errors.Errorf(
+			"transform middleware found with no options.",
+		)
+	}
+	transforms := opts["transforms"].([]interface{})
```
Comment (cl1337) [inline codegen/gateway.go]: fix casting `[]map[string]interface{}`

**Comment by Raynos** [comment]
Great, we have tests for the backend and the code generation.

If possible can we also add tests to `test/endpoints/baz/...` to verify that sending HTTP requests works as expected ?

**Comment by cl1337** [comment]
@Raynos I tried but our test template is broken, for HTTP response we're asserting equal JSON bytes, which, will not guarantee test pass every generate, for example:
we generate: 
```
	assert.Equal(t, []byte(`{"driver":{"check":12,"msg":"tchan_return_driver"},"message":"tchan_return_msg","rider":{"check":11,"msg":"tchan_return_msg"}}`), data)
```
with result failing:
```
	            	Diff:
	            	--- Expected
	            	+++ Actual
	            	@@ -1,10 +1,10 @@
	            	 ([]uint8) (len=126) {
	            	- 00000000  7b 22 64 72 69 76 65 72  22 3a 7b 22 63 68 65 63  |{"driver":{"chec|
	            	- 00000010  6b 22 3a 31 32 2c 22 6d  73 67 22 3a 22 74 63 68  |k":12,"msg":"tch|
	            	- 00000020  61 6e 5f 72 65 74 75 72  6e 5f 64 72 69 76 65 72  |an_return_driver|
	            	- 00000030  22 7d 2c 22 6d 65 73 73  61 67 65 22 3a 22 74 63  |"},"message":"tc|
	            	- 00000040  68 61 6e 5f 72 65 74 75  72 6e 5f 6d 73 67 22 2c  |han_return_msg",|
	            	- 00000050  22 72 69 64 65 72 22 3a  7b 22 63 68 65 63 6b 22  |"rider":{"check"|
	            	- 00000060  3a 31 31 2c 22 6d 73 67  22 3a 22 74 63 68 61 6e  |:11,"msg":"tchan|
	            	- 00000070  5f 72 65 74 75 72 6e 5f  6d 73 67 22 7d 7d        |_return_msg"}}|
	            	+ 00000000  7b 22 6d 65 73 73 61 67  65 22 3a 22 74 63 68 61  |{"message":"tcha|
	            	+ 00000010  6e 5f 72 65 74 75 72 6e  5f 6d 73 67 22 2c 22 64  |n_return_msg","d|
	            	+ 00000020  72 69 76 65 72 22 3a 7b  22 6d 73 67 22 3a 22 74  |river":{"msg":"t|
	            	+ 00000030  63 68 61 6e 5f 72 65 74  75 72 6e 5f 64 72 69 76  |chan_return_driv|
	            	+ 00000040  65 72 22 2c 22 63 68 65  63 6b 22 3a 31 32 7d 2c  |er","check":12},|
	            	+ 00000050  22 72 69 64 65 72 22 3a  7b 22 6d 73 67 22 3a 22  |"rider":{"msg":"|
	            	+ 00000060  74 63 68 61 6e 5f 72 65  74 75 72 6e 5f 6d 73 67  |tchan_return_msg|
	            	+ 00000070  22 2c 22 63 68 65 63 6b  22 3a 31 31 7d 7d        |","check":11}}|
	            	 }
```

**Comment by cl1337** [comment]
for the record, I'm adding this test case once I fixed the broken test template:
```
{
	"endpointId": "baz",
	"endpointType": "http",
	"handleId": "trans",
	"thriftFile": "endpoints/baz/baz.thrift",
	"thriftFileSha": "{{placeholder}}",
	"thriftMethodName": "SimpleService::trans",
	"workflowType": "tchannelClient",
	"clientId": "baz",
	"clientMethod": "Trans",
	"middlewares": [
		{
			"name": "transformRequest",
			"options": {
				"transforms": [
					{
						"from": "Arg1.Driver.Msg",
						"to": "Arg2.Rider.Msg",
						"override": true
					},
					{
						"from": "Arg1.Rider",
						"to": "Arg2.Driver",
						"override": false
					}
				]
			}
		},
		{
			"name": "transformResponse",
			"options": {
				"transforms": [
					{
						"from": "Message",
						"to": "Rider.Msg",
						"override": true
					},
					{
						"from": "Message",
						"to": "Driver.Msg",
						"override": false
					}
				]
			}
		}
	],
	"reqHeaderMap": {},
	"resHeaderMap": {},
	"testFixtures": {
		"successfulRequest": {
			"testName": "successfulRequest",
			"endpointId": "baz",
			"handleId": "trans",
			"endpointRequest": {
				"requestType": "http",
				"httpRequest": {
					"method": "POST",
					"body": {
						"bodyType": "json",
						"bodyJson": {
							"arg1": {
								"message": "msg_arg1",
								"rider": {
									"msg": "arg1_rider_msg",
									"check": 1
								},
								"driver": {
									"msg": "arg1_driver_msg",
									"check": 2
								}
							},
							"arg2": {
								"message": "msg_arg2",
								"rider": {
									"msg": "arg2_rider_msg",
									"check": 3
								},
								"driver": {
									"msg": "arg2_driver_msg",
									"check": 4
								}
							},
							"message": "message"
						}
					}
				}
			},
			"endpointReqHeaders": {},
			"endpointResponse": {
				"responseType": "http",
				"httpResponse": {
					"statusCode": 200,
					"body": {
						"bodyType": "json",
						"bodyJson": {
							"message": "tchan_return_msg",
							"rider": {
								"msg": "tchan_return_msg",
								"check": 11
							},
							"driver": {
								"msg": "tchan_return_driver",
								"check": 12
							}
						}
					}
				}
			},
			"endpointResHeaders": {},
			"clientTestFixtures": {
				"baz::call": {
					"clientId": "baz",
					"clientMethod": "trans",
					"clientRequest": {
						"requestType": "tchannel",
						"tchannelRequest": {
							"arg1": {
								"message": "msg_arg1",
								"rider": {
									"msg": "arg1_rider_msg",
									"check": 1
								},
								"driver": {
									"msg": "arg1_driver_msg",
									"check": 2
								}
							},
							"arg2": {
								"message": "msg_arg2",
								"rider": {
									"msg": "arg1_driver_msg",
									"check": 3
								},
								"driver": {
									"msg": "arg2_driver_msg",
									"check": 4
								}
							},
							"message": "message"
						}
					},
					"clientReqHeaders": {},
					"clientResponse": {
						"responseType": "tchannel",
						"tchannelResponse": {
							"message": "tchan_return_msg",
							"rider": {
								"msg": "tchan_return_rider",
								"check": 11
							},
							"driver": {
								"msg": "tchan_return_driver",
								"check": 12
							}
						}
					},
					"clientResHeaders": {}
				}
			},
			"testServiceName": "example-gateway"
		}
	}
}

```

**Pull request merged**